### PR TITLE
Fix categorical crossentropy and Masking

### LINKED ISSUE #1567: Does Masking have a bug?
It seems that `K.rnn` thinks that mask has 3 dimensions. How to get the error:

```python
from keras.models import  Sequential
from keras.layers.recurrent import LSTM
from keras.layers.core import Masking

model = Sequential()
model.add(Masking(input_shape=(None, 8)))
model.add(LSTM(output_dim=256, return_sequences=True))

model.compile(loss='binary_crossentropy', optimizer='adam')
```

And the error is:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-23d1674d2143> in <module>()
----> 1 model.compile(loss='binary_crossentropy', optimizer='adam')

/home/eders/python/keras/keras/models.pyc in compile(self, optimizer, loss, class_mode)
    433         self.X_test = self.get_input(train=False)
    434 
--> 435         self.y_train = self.get_output(train=True)
    436         self.y_test = self.get_output(train=False)
    437 

/home/eders/python/keras/keras/layers/containers.pyc in get_output(self, train)
    126 
    127     def get_output(self, train=False):
--> 128         return self.layers[-1].get_output(train)
    129 
    130     def set_input(self):

/home/eders/python/keras/keras/layers/recurrent.pyc in get_output(self, train)
    148                                              initial_states,
    149                                              go_backwards=self.go_backwards,
--> 150                                              mask=mask)
    151         if self.stateful:
    152             self.updates = []

/home/eders/python/keras/keras/backend/theano_backend.pyc in rnn(step_function, inputs, initial_states, go_backwards, mask)
    448         mask = expand_dims(ones_like(T.sum(inputs, axis=-1)))
    449     else:
--> 450         mask = mask.dimshuffle(axes)
    451 
    452     def _step(input, mask, output_tm1, *states):

/home/eders/python/Theano/theano/tensor/var.pyc in dimshuffle(self, *pattern)
    358             pattern = pattern[0]
    359         op = theano.tensor.basic.DimShuffle(list(self.type.broadcastable),
--> 360                                             pattern)
    361         return op(self)
    362 

/home/eders/python/Theano/theano/tensor/elemwise.pyc in __init__(self, input_broadcastable, new_order, inplace)
    144                     raise ValueError(("new_order[%d] is %d, but the input "
    145                                       "only has %d axes.") %
--> 146                                      (i, j, len(input_broadcastable)))
    147                 if j in new_order[(i + 1):]:
    148                     raise ValueError("The same input dimension may not appear "

ValueError: new_order[2] is 2, but the input only has 2 axes.
```

### PR DESCRIPTION
**Pull request by wxs**
Fixes #1567

- The RNN backend call can now take a mask with or without the same number of dimensions as the input data
- As part of the investigation, I found that categorical_crossentropy was taking an extra mean, the function already removes the final dimension of your input, so you don't need to take a mean as you would with, say, L2 loss. This seems like a pretty severe potential issue so anyone doing sequence to sequence learning should take note.

- Since I was already down that rabbit-hole, I also fixed Masking to work with tensorflow.

### CODE REVIEW

File: keras/layers/core.py [hunk 1]
```diff
@@ -319,17 +319,16 @@ class Masking(MaskedLayer):
     def __init__(self, mask_value=0., **kwargs):
         super(Masking, self).__init__(**kwargs)
         self.mask_value = mask_value
-        self.input = K.placeholder(ndim=3)
+        if (not hasattr(self, 'input')):
```
Comment (fchollet) [inline keras/layers/core.py]: Would you expect `input` to be set by the parent class? Why?

Comment (wxs) [inline keras/layers/core.py]: It is set by the parent class when you specify the `input_shape`, so us overriding it afterwards was causing issues in some tests.

**Comment by wxs** [comment]
FYI I just added some tests to make sure that all the current objective functions have the desired output shape in 2d and 3d.

File: keras/objectives.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def hinge(y_true, y_pred):
 def categorical_crossentropy(y_true, y_pred):
     '''Expects a binary class matrix instead of a vector of scalar classes.
     '''
-    return K.mean(K.categorical_crossentropy(y_pred, y_true), axis=-1)
+    return K.categorical_crossentropy(y_pred, y_true)
```
Comment (fchollet) [inline keras/objectives.py]: Right, I didn't catch this.

File: keras/objectives.py [hunk 3]
```diff
@@ -49,11 +49,9 @@ def poisson(y_true, y_pred):
...
 
 def cosine_proximity(y_true, y_pred):
-    assert K.ndim(y_true) == 2
-    assert K.ndim(y_pred) == 2
-    y_true = K.l2_normalize(y_true, axis=1)
-    y_pred = K.l2_normalize(y_pred, axis=1)
-    return -K.mean(y_true * y_pred, axis=1)
+    y_true = K.l2_normalize(y_true, axis=-1)
+    y_pred = K.l2_normalize(y_pred, axis=-1)
+    return -K.mean(y_true * y_pred, axis=-1)
```
Comment (fchollet) [inline keras/objectives.py]: Time distributed cosine proximity? Is this something you are using?

File: tests/keras/layers/test_recurrent.py [hunk 4]
```diff
@@ -100,5 +103,20 @@ def test_LSTM():
...
+def test_masking_layer():
+    ''' This test based on a previously failing issue here:
+    https://github.com/fchollet/keras/issues/1567
+
+    '''
+    model = Sequential()
+    model.add(Masking(input_shape=(12, 8)))
+    model.add(recurrent.LSTM(output_dim=256, return_sequences=True))
+    model.compile(loss='categorical_crossentropy', optimizer='adam')
+    I = np.random.random((10,12,8))
```
Comment (fchollet) [inline tests/keras/layers/test_recurrent.py]: Space after `,` (if only for consistency with the rest of the function).

**Comment by fchollet** [comment]
Looks good to me.

File: keras/objectives.py (hunk 3 above)
Comment (wxs) [inline keras/objectives.py]: No I'm not, it just came up while I was writing tests that it didn't support time-distributed and I figured there was no downside to adding it and making it consistent with the other objectives.

**Comment by fchollet** [comment]
Looking good. Please fix the remaining PEP8 issue, squash your commits, and I will merge.

**Comment by wxs** [comment]
OK after some messing around to resolve merge conflicts I've got a finalized PR. FYI I made one small change to the shape of the mask coming out of RNN layers. Theano was having trouble correctly broadcasting for reasons that are not completely clear.

**Pull request merged**